feat: GHE support for Go releases

partha04patel opened the pull request:
Fixes #664

mrgrain commented on src/help/git.ts:
```diff
@@ -13,10 +13,19 @@ export function clone(repositoryUrl: string, targetDir: string) {
     const gitHubToken = process.env.GITHUB_TOKEN;
-    if (!gitHubToken) {
-      throw new Error('GITHUB_TOKEN env variable is required when GITHUB_USE_SSH env variable is not used');
+    if (gitHubToken) {
+      shell.run(`git clone https://${gitHubToken}@${repositoryUrl}.git ${targetDir}`);
+    } else {
```
If I read this correctly `GITHUB_TOKEN` is present even in GHE workflows: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

mrgrain replied:
So I think the logic goes something like this

```
function getToken(isGHE: boolean) {
  if (isGhE()) {
    return process.env.GH_ENTERPRISE_TOKEN ?? process.env.GITHUB_TOKEN; 
  }
  return process.env.GITHUB_TOKEN;
}
```

```
const isGHE = detectGHE();
const token = getToken(isGHE);
if (!token) throw Error;
shell.run(`git clone https://${token}@${repositoryUrl}.git ${targetDir}`);
```

mrgrain reviewed (changes requested):
Some logic improvements needed and also can you please update the docs (once we have settled on the API)

partha04patel commented:
I separated some of the logic into helper functions, for docs, should I be changing any of the md files like the README? I wasn't sure if that was autogenerated by CI or something, but I can add some notes there. I think one thing that may need to be called out maybe as a warning or something is that if someone was running a go release with publib in something that's not GHA, they'd need the GITHUB_API_URL defined since that's how the logic would detect if the env is either public GitHub or GHE, and that's a default env var that's present in a GHA workflow.

mrgrain commented:
Hadn't had time to look at the updated code yet, but I can answer the question.


> For docs, should I be changing any of the md files like the README? I added docstrings to the helper functions but wasn't sure if the README/other doc files were autogenerated by CI or something, but I can add some notes there if it's not. 

Yes, please update the goland section here: https://github.com/cdklabs/publib?tab=readme-ov-file#golang
Especially the env var table.

> I think one thing that may need to be called out maybe as a warning or something is that if someone was running a go release with publib in something that's not GHA, they'd need the GITHUB_API_URL defined since that's how the logic would detect if the env is either public GitHub or GHE, and that's a default env var that's present in a GHA workflow.

We can support both `GITHUB_API_URL` and `GH_HOST` (from the [CLI conventions](https://cli.github.com/manual/gh_help_environment)). But definitely needs to be called out in the docs.

partha04patel commented:
Added some extra documentation to the README!

> Hadn't had time to look at the updated code yet, but I can answer the question.
> 
> > For docs, should I be changing any of the md files like the README? I added docstrings to the helper functions but wasn't sure if the README/other doc files were autogenerated by CI or something, but I can add some notes there if it's not.
> 
> Yes, please update the goland section here: https://github.com/cdklabs/publib?tab=readme-ov-file#golang Especially the env var table.
> 
> > I think one thing that may need to be called out maybe as a warning or something is that if someone was running a go release with publib in something that's not GHA, they'd need the GITHUB_API_URL defined since that's how the logic would detect if the env is either public GitHub or GHE, and that's a default env var that's present in a GHA workflow.
> 
> We can support both `GITHUB_API_URL` and `GH_HOST` (from the [CLI conventions](https://cli.github.com/manual/gh_help_environment)). But definitely needs to be called out in the docs.

mrgrain commented on README.md:
```diff
@@ -221,8 +221,11 @@ Repository tags will be in the following format:
-|`GITHUB_TOKEN`|Required if `GITHUB_USE_SSH` is not specified|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
-|`GITHUB_USE_SSH`|Required if `GITHUB_TOKEN` is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
+|`GITHUB_TOKEN`|Required if either `GITHUB_USE_SSH`\\`GIT_USE_SSH` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST` is not specified|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
+|`GITHUB_USE_SSH` or `GIT_USE_SSH`|Required if `GITHUB_TOKEN` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST`is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
```
Let's align on `GIT_USE_SSH` for most of the doc.
```suggestion
|`GIT_USE_SSH`|Required if `GITHUB_TOKEN` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST`is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
|`GITHUB_USE_SSH`|Deprecated. Alias for `GIT_USE_SSH`.|
```

partha04patel replied:
sounds good to me, committed!

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+ * @return either GH_ENTERPRISE_TOKEN or GITHUB_ENTERPRISE_TOKEN if one of the two and the GH_HOST env var is set, otherwise returns GITHUB_TOKEN env var.
+ */
+
+export function getToken(isGHE: boolean): (string | undefined) {
```
```suggestion
function getToken(isGHE: boolean): (string | undefined) {
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+ * @return GIT_USE_SSH env var if it's defined, otherwise returns GITHUB_USE_SSH env var.
+ */
+
+export function detectSSH(): (string | undefined) {
+  const gitHubUseSsh = process.env.GIT_USE_SSH ?? process.env.GITHUB_USE_SSH;
+  return gitHubUseSsh;
```
```suggestion
export function detectSSH(): boolean {
  return Boolean(process.env.GIT_USE_SSH ?? process.env.GITHUB_USE_SSH);
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+}
+
+/**
+ * Checks for the presence of SSH-related env vars in the GitHub Actions workflow to see if SSH should be used to clone repo.
+ * @return GIT_USE_SSH env var if it's defined, otherwise returns GITHUB_USE_SSH env var.
```
```suggestion
 * Checks if SSH should be used to clone repo. 
 * This checks the presence and values of the GIT_USE_SSH env variable and the deprecated GITHUB_USE_SSH for legacy reason. Returns true if either of these env vars are defined and not falsey.
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+ *
+ * @return True if GITHUB_API_URL env var is defined and not equal to public github API URL, false,
+ */
+export function detectGHE(): boolean {
```
Let's be nice and also check `process.env.GH_HOST` in here:

```
(GH_HOST && GH_HOST != 'github.com') || GITHUB_API_URL && GITHUB_API_URL != 'https://api.github.com'
```

each with the lowercase trim

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+ * @return True if GITHUB_API_URL env var is defined and not equal to public github API URL, false,
+ */
+export function detectGHE(): boolean {
+  const githubApiUrl = process.env.GITHUB_API_URL;
+  if (!githubApiUrl) {
+    return false;
+  }
+  return githubApiUrl!.trim().toLowerCase()!= 'https://api.github.com';
+
```
```suggestion
  const githubApiUrl = process.env.GITHUB_API_URL;
  const ghHost = process.env.GH_HOST;
  
  return (ghHost && ghHost!.trim().toLowerCase() != 'github.com')
    || (githubApiUrl && githubApiUrl!.trim().toLowerCase() != 'https://api.github.com')
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+export function getToken(isGHE: boolean): (string | undefined) {
+  if (isGHE) {
+    const githubEnterpiseToken = process.env.GH_ENTERPRISE_TOKEN ?? process.env.GITHUB_ENTERPRISE_TOKEN;
+    const githubEnterpriseHost = process.env.GH_HOST;
```
This check seems unnecessary here

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+export function getToken(isGHE: boolean): (string | undefined) {
+  if (isGHE) {
+    const githubEnterpiseToken = process.env.GH_ENTERPRISE_TOKEN ?? process.env.GITHUB_ENTERPRISE_TOKEN;
+    const githubEnterpriseHost = process.env.GH_HOST;
+    if (githubEnterpiseToken && githubEnterpriseHost) {
+      return githubEnterpiseToken;
+    }
```
```suggestion
    return githubEnterpiseToken;
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+
+export function getToken(isGHE: boolean): (string | undefined) {
+  if (isGHE) {
+    const githubEnterpiseToken = process.env.GH_ENTERPRISE_TOKEN ?? process.env.GITHUB_ENTERPRISE_TOKEN;
```
Pretty sure on GH Actions it's just the same token:
```suggestion
    const githubEnterpiseToken = process.env.GH_ENTERPRISE_TOKEN ?? process.env.GITHUB_ENTERPRISE_TOKEN ?? process.env. GITHUB_TOKEN;
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+}
+
+/**
+ * Checks for the presence of GITHUB_TOKEN set in the GitHub Actions workflow, as well as if there are GHE related env vars set.
+ * https://cli.github.com/manual/gh_help_environment
+ *
+ * @return either GH_ENTERPRISE_TOKEN or GITHUB_ENTERPRISE_TOKEN if one of the two and the GH_HOST env var is set, otherwise returns GITHUB_TOKEN env var.
```
```suggestion
 * Returns an appropriate github token from the environment.
 *
 * @return GH_ENTERPRISE_TOKEN or GITHUB_ENTERPRISE_TOKEN or GITHUB_TOKEN if in an GHE environment, otherwise GITHUB_TOKEN
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,61 @@ import * as shell from './shell';
+}
+
+/**
+ * Checks if the GitHub API URL set in the GitHub Actions workflow is not the public GitHub API URL.
+ * https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
+ *
+ * @return True if GITHUB_API_URL env var is defined and not equal to public github API URL, false,
```
```suggestion
 * Checks if the current environment is an GHE environment.
 *
 * This check is using GITHUB_API_URL set in GitHub Actions workflow, as well as common gh cli env variables.
 * https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
 * https://cli.github.com/manual/gh_help_environment
 *
 * @return - `true` if GH_HOST or GITHUB_API_URL env var are defined and not equal to the public github endpoint, otherwise `false`
```

mrgrain commented on README.md:
```diff
@@ -221,8 +221,12 @@ Repository tags will be in the following format:
+|`GIT_USE_SSH`|Required if `GITHUB_TOKEN` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST`is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
+|`GITHUB_USE_SSH`| Deprecated | Legacy alias for `GIT_USE_SSH`.|
+|`GH_ENTERPRISE_TOKEN` or  `GITHUB_ENTERPRISE_TOKEN`|Required if `GITHUB_TOKEN` and `GITHUB_USE_SSH`\\`GIT_USE_SSH`is not specified|[Authenticationt token for API requests to GitHub Entperise](https://cli.github.com/manual/gh_help_environment).|
+|`GH_HOST`|Required if either `GITHUB_ENTERPRISE_TOKEN` or `GH_ENTPERISE_TOKEN` is specified|[Hostname for GitHub Enterprise](https://cli.github.com/manual/gh_help_environment).|
```
```suggestion
|`GH_HOST`| Optional | Force use of a different [Hostname for GitHub Enterprise](https://cli.github.com/manual/gh_help_environment).|
```

mrgrain commented on README.md:
```diff
@@ -221,8 +221,12 @@ Repository tags will be in the following format:
+|`GITHUB_USE_SSH`| Deprecated | Legacy alias for `GIT_USE_SSH`.|
+|`GH_ENTERPRISE_TOKEN` or  `GITHUB_ENTERPRISE_TOKEN`|Required if `GITHUB_TOKEN` and `GITHUB_USE_SSH`\\`GIT_USE_SSH`is not specified|[Authenticationt token for API requests to GitHub Entperise](https://cli.github.com/manual/gh_help_environment).|
+|`GH_HOST`|Required if either `GITHUB_ENTERPRISE_TOKEN` or `GH_ENTPERISE_TOKEN` is specified|[Hostname for GitHub Enterprise](https://cli.github.com/manual/gh_help_environment).|
+|`GITHUB_API_URL`|Optional| API URL for instance of GitHub to target. This is specified by default in [GitHub Actions workflow](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables).|
```
```suggestion
|`GITHUB_API_URL`| Optional | If present, used to detect the GitHub instance to target. This is specified by default in [GitHub Actions workflow](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables) and should not be set manually.|
```

mrgrain commented on README.md:
```diff
@@ -221,8 +221,12 @@ Repository tags will be in the following format:
-|`GITHUB_TOKEN`|Required if `GITHUB_USE_SSH` is not specified|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
-|`GITHUB_USE_SSH`|Required if `GITHUB_TOKEN` is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
+|`GITHUB_TOKEN`|Required if either `GITHUB_USE_SSH`\\`GIT_USE_SSH` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST` is not specified|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
+|`GIT_USE_SSH`|Required if `GITHUB_TOKEN` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST`is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
```
```suggestion
|`GIT_USE_SSH`|Required if in SSH mode|Set to a non-falsy value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
```

mrgrain commented on README.md:
```diff
@@ -221,8 +221,12 @@ Repository tags will be in the following format:
 |------|--------|-----------|
-|`GITHUB_TOKEN`|Required if `GITHUB_USE_SSH` is not specified|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
-|`GITHUB_USE_SSH`|Required if `GITHUB_TOKEN` is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
+|`GITHUB_TOKEN`|Required if either `GITHUB_USE_SSH`\\`GIT_USE_SSH` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST` is not specified|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
```
```suggestion
|`GITHUB_TOKEN`|Required when not using SSH mode, see `GIT_USE_SSH`|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
```

mrgrain commented on README.md:
```diff
@@ -221,8 +221,12 @@ Repository tags will be in the following format:
+|`GITHUB_TOKEN`|Required if either `GITHUB_USE_SSH`\\`GIT_USE_SSH` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST` is not specified|[GitHub personal access token.](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)|
+|`GIT_USE_SSH`|Required if `GITHUB_TOKEN` and combination of `GH_ENTERPRISE_TOKEN`\\`GITHUB_ENTERPISE_TOKEN` and `GH_HOST`is not specified|Set any value to use SSH with [deploy keys](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) or your private SSH key. Your system must ready to use the key as publib will not set it up.|
+|`GITHUB_USE_SSH`| Deprecated | Legacy alias for `GIT_USE_SSH`.|
+|`GH_ENTERPRISE_TOKEN` or  `GITHUB_ENTERPRISE_TOKEN`|Required if `GITHUB_TOKEN` and `GITHUB_USE_SSH`\\`GIT_USE_SSH`is not specified|[Authenticationt token for API requests to GitHub Entperise](https://cli.github.com/manual/gh_help_environment).|
```
```suggestion
|`GH_ENTERPRISE_TOKEN`|Optional|[Custom Authentication token for API requests to GitHub Enterprise](https://cli.github.com/manual/gh_help_environment).|
|`GITHUB_ENTERPRISE_TOKEN`|Optional|[Custom Authentication token for API requests to GitHub Enterprise](https://cli.github.com/manual/gh_help_environment).|
```

mrgrain commented on src/help/git.ts:
```diff
@@ -7,17 +7,58 @@ import * as shell from './shell';
+
+export function getToken(isGHE: boolean): (string | undefined) {
+  if (isGHE) {
+    const githubEnterpiseToken = process.env.GH_ENTERPRISE_TOKEN ?? process.env.GITHUB_ENTERPRISE_TOKEN ?? process.env. GITHUB_TOKEN;
```
```suggestion
    const githubEnterpiseToken = process.env.GH_ENTERPRISE_TOKEN ?? process.env.GITHUB_ENTERPRISE_TOKEN ?? process.env.GITHUB_TOKEN;
```
